Do not bind Arel attributes

chewi opened the pull request:
This (slightly convoluted example) used to work but no longer does.

``` ruby
Company.where(:name => Company.arel_table[:name])
```

This is especially useful when eager loading associations with scopes containing additional where conditions.

Schwad commented:
@chewi  I see in your update on your comment on https://github.com/rails/rails/issues/15024 you feel that your approach suggested here has been rejected elsewhere. Do you still feel that way or what's your current perspective here?

@arthurnn do you have any general thoughts on this PR?

chewi commented:
@Schwad, that comment was about the issue in #15024 and nothing to do with this. I would still very much like this to be merged.

rafaelfranca commented:
Usually the answer for this kind of issue is: Arel is private API of Rails and it is not recommended to be used in applications. But I think we made part of arel public in Rails 5.2.

@matthewd what do you think about this?

chewi committed:
Do not bind Arel attributes

This (slightly convoluted example) used to work but no longer does.

  Company.where(:name => Company.arel_table[:name])

This is especially useful when eager loading associations with scopes
containing additional where conditions.

chewi commented:
Still an issue. Rebased.

kamipo commented on activerecord/test/cases/finder_test.rb:
```diff
@@ -1001,6 +1001,12 @@ def test_hash_condition_local_time_interpolation_with_default_timezone_utc
   end
 
+  def test_condition_arel_attribute
+    relation = Company.where(name: Company.arel_table[:name])
```
Why not use `where(Company.arel_table[:name].eq(Company.arel_table[:name]))`?

chewi replied:
I've since left the company where I was using this although they are still relying on it and I could see myself needing it again. I must admit I can't remember why I didn't do that instead but I am aware that it's possible so there probably was a reason, perhaps that it's not possible in some other contexts? It's a very small but useful change in any case.

kamipo replied:
Adding extra predicate handler makes building all predicate slower.
If `where(Company.arel_table[:name].eq(Company.arel_table[:name]))` is a common use case, people will accept to be slower.
But I don't think that is a common use case, I'm not prefer to be slower by default for making that.

chewi replied:
Fair enough.
